### PR TITLE
Refactor project 

### PR DESCRIPTION
**Pull request by lilliCao**
Refactor project structure and naming convention after [this schema](https://github.com/hohwille/devon4j/blob/feature/420-d4q-coding/documentation/guide-structure-modern.asciidoc#architecture-mapping)
* rest.v1 -> service.v1
* RestController -> RestService
* Animal -> AnimalEntity

### CODE REVIEW

File: src/main/java/com/devonfw/demoquarkus/service/AnimalRestService.java [hunk 1]
```diff
@@ -1,4 +1,4 @@
-package com.devonfw.demoquarkus.rest.v1.controller;
+package com.devonfw.demoquarkus.service;
```
Comment (GuentherJulian) [inline src/main/java/com/devonfw/demoquarkus/service/AnimalRestService.java:1]: Wouldn't it make sense to keep the version information (`com.devonfw.demoquarkus.service.v1`) as described here  https://github.com/devonfw/devon4j/blob/master/documentation/guide-service-layer.asciidoc#versioning?

Comment (lilliCao) [reply src/main/java/com/devonfw/demoquarkus/service/AnimalRestService.java:1]: I think we should keep it simple ( without a version like other samples mtmj, jtq) unless we want to demonstrate the versioning. @hohwille any thought?

**Review by GuentherJulian — APPROVED** [review]

**Review by hohwille — COMMENTED** [review]
@lilliCao thanks for this PR. I see two issues to be discussed:
* When we do not follow the Entity / ETO approach of devonfw, we should clarify if we want to enforce the `Entity` suffix or if we just leave this as is. I personally like if types have a name that (wihtout the package) already tells what it is. Also it would make our lives a lot easier with CobiGen, etc. if we can keep this suffix. But let us first get some more feedback here...
* For the service you also dropped the versioning from the package. There is an ongoing discussion to create a new service by "forking" the app and then having both versions (e.g. v1 and v2) as separate apps, containers and processes. In that case it would be a nice simplification (at least for dev maybe not for deployment and SCM). Also in devon4j we so far only prepared the structure and design for service versioning but tried to keep the default as simple as possible. So the package without any version is the latest version. Only when you need to start changing an API incompatible (and have consumers that can not be changed at the same time), we copy the existing code to a `v«x»` package and change the "latest" API afterwards. Therefore IMHO this change is fine as well.

renaming of package and classes for service is fine and as suggested by @maihacke 
Looks good to me.

So from my PoV I am happy to merge. However, my intention is also to keep others involved and hear their pros and cons.
Any feedback from OCX?

**Comment by capap** [comment]
@lilliCao @hohwille 


> * For the service you also dropped the versioning from the package. There is an ongoing discussion to create a new service by "forking" the app and then having both versions (e.g. v1 and v2) as separate apps, containers and processes. In that case it would be a nice simplification (at least for dev maybe not for deployment and SCM). Also in devon4j we so far only prepared the structure and design for service versioning but tried to keep the default as simple as possible. So the package without any version is the latest version. Only when you need to start changing an API incompatible (and have consumers that can not be changed at the same time), we copy the existing code to a `v«x»` package and change the "latest" API afterwards. Therefore IMHO this change is fine as well.


Packages for versioning is one side of the problem, but we need to keep in mind also the 'path' and `openapi` schema.
When we have the latest and no package versioning, we need to keep it without change and a make new version with a new path otherwise, it will break contract to the consumer and API cannot be classified as stable.
We can still have no version in package and version in path `/v1/* but this could be confusing for developers. For the API-First approach, developers do not have a control over the versioning of the API. In few services, new version v2 was already planned and designed before the version v1 was in production.


Making a new version with separated microservice will work for simple microservice and where you do not need to support multiple versions of the API for long time period. With this approach, we need to copy also the business logic and support multiple source bases. It is not possible, but we should have written down the downside of this approach.

Andrej

**Review by hohwille — CHANGES_REQUESTED** [review]
@capap Thanks for your feedback. I agree.

> Packages for versioning is one side of the problem, but we need to keep in mind also the 'path' and openapi schema.

Sure. The path was also not changed by @lilliCao.

> When we have the latest and no package versioning, we need to keep it without change and a make new version with a new path otherwise, it will break contract to the consumer and API cannot be classified as stable.

Exactly. We have an existing guide for long already and I think we all agree on the same concept:
https://github.com/devonfw/devon4j/blob/master/documentation/guide-service-versioning.asciidoc
IMHO the guide is still valid but the last two sections (starting with `Practice`) will need a facelift for our quarkus, cloud & microservice initiative we are currently all working on.

> We can still have no version in package and version in path `/v1/*` but this could be confusing for developers.

I see that point and therefore also raised this review concern.
So far we used to do exactly this in devon4j and it worked out fine for developers.
Howevere, maybe that was because only very few projects actually later ran into the situation that they had to support multiple versions in parallel. This is always kind of complex and conusing even independent of the packaging conventions...
In the end both ways are fine for me. We should only agree on what we do for our sample apps as reference showcase.

@lilliCao could you please rework this PR and keep the `v1` currently in the pacakging. Then we (e.g. Jan-Gerrit as I will go on vacation) can merge as the rest seems to be fine. Then please also raise a backlog issue about this discussion here so that we still can track it.

> For the API-First approach, developers do not have a control over the versioning of the API. 

Valid point. If you do code-generation you most probably can not even control the packaging much.
Maybe we should also create a separate minimum sample app with quarkus that shows a rest service where the boilerplate code is generated from openAPI as this is also an important use-case. From the output of that generation we can revisit our recommendation.

> In few services, new version v2 was already planned and designed before the version v1 was in production.

Also important to explain that service-versioning is causing real effort and when this effort can be avoided at all because maybe in that situation breaking v1 temporary in a test-environment is not causing big pain.
I once tried to express this in the guide linked above. Everybody is invited to improve that guide.

> Making a new version with separated microservice will work for simple microservice and where you do not need to support multiple versions of the API for long time period. With this approach, we need to copy also the business logic and support multiple source bases. It is not possible, but we should have written down the downside of this approach.

I also agree. Maybe it is even better to let the projects decide themselves. There might be even other options, we are not seeing. Following our solid principle of devonfw we can wait until projects really found out a better way to do this that is not a theoretical idea but proven to work better in practice. Then we can still add such suggestion as option with pros and cons to our docs.

**Review by hohwille — APPROVED** [review]
@lilliCao excellent. Thanks for the rework. 👍 
I will now merge this PR.

**Pull request merged**